src/main/java/jenkins/scm/api/SCMSource.java

```diff
@@ -956,13 +958,37 @@ public final SCM build(@NonNull SCMHead head) {
      * @throws IOException in case the implementation must call {@link #fetch(SCMHead, TaskListener)} or similar
      * @throws InterruptedException in case the implementation must call {@link #fetch(SCMHead, TaskListener)} or similar
      * @since 1.1
+     * @see #getTrustedRevisionForBuild
      */
     @NonNull
     public SCMRevision getTrustedRevision(@NonNull SCMRevision revision, @NonNull TaskListener listener)
             throws IOException, InterruptedException {
         return revision;
     }
 
+    /**
+     * Refined version of {@link #getTrustedRevision(SCMRevision, TaskListener)} that takes into account the build context.
+     * @param build a running build
+     * @return {@link #getTrustedRevision} if the build itself does not indicate trust;
+     *         simply {@code revision} if any {@link TrustworthyBuild} says that it does
+     */
+    @NonNull
+    public final SCMRevision getTrustedRevisionForBuild(@NonNull SCMRevision revision, @NonNull TaskListener listener, @NonNull Run<?, ?> build)
+            throws IOException, InterruptedException {
+        if (ExtensionList.lookup(TrustworthyBuild.class).stream().anyMatch(tb -> tb.shouldBeTrusted(build, listener))) {
+            LOGGER.fine(() -> build + " with " + build.getCauses() + " was considered trustworthy, so using " + revision + " as is");
+            return revision;
+        } else {
+            SCMRevision trustedRevision = getTrustedRevision(revision, listener);
+            if (trustedRevision.equals(revision)) {
+                LOGGER.fine(() -> build + " was not considered trustworthy, but " + revision + " was trusted anyway");
```

Comment:
This seems misleading. Many builds will fall into this category, as their build "causes" will not result in trust, but the revision itself will be trusted.

it seems more natural to me to first check for `#getTrustedRevision` and only check trustworthiness if that is different from the actual `revision`, but I guess it depends on what we consider cheaper to compute?

Reply:
`TrustworthyBuild` should be cheaper to compute than `getTrustedRevision`, so I would leave the logic in this order and just clarify the log messages.